File: HW/HW9/solution.java
package HW.HW9;

public class solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {

        //These 2 If statments cover the base cases of the algorithm

        if ((p.val < root.val && q.val > root.val) || (p.val > root.val && q.val < root.val)){
            System.out.println("BASE CASE 2 REACHED");
            return root;
        }

        if(root.val == q.val || root.val == p.val){
            System.out.println("Comparison 1 REACHED");
            return root;
        }
        //This part of the code provides how to traverse recursively.
        if(p.val < root.val && q.val < root.val){
            System.out.println("Comparison 2 REACHED");
            return lowestCommonAncestor(root.left, p, q);
        }
        else if(p.val > root.val && q.val > root.val){
            System.out.println("Comparison 3 REACHED");
            return lowestCommonAncestor(root.right, p, q);
        }

        return null;

        
     }

     

     public static void main(String[] args) {

        //Test Case you provided
        TreeNode root = new TreeNode(4);
        TreeNode p = root.left = new TreeNode(3);
        root.right = new TreeNode(8);
        TreeNode q = root.left.left = new TreeNode(1);
        //root.left.right = new TreeNode(5);
        root.right.left = new TreeNode(5);
        root.right.right = new TreeNode(9);
        
        solution test1 = new solution();
        
        TreeNode solution = test1.lowestCommonAncestor(root, p, q);
        System.out.println(solution.val);

        //Test Case where p and q are directly under the root
        TreeNode root3 = new TreeNode(4);
        TreeNode p3 = root3.left = new TreeNode(5);
        TreeNode q3 = root3.right = new TreeNode(1);   
        
        solution test3 = new solution();
        
        TreeNode solution3 = test3.lowestCommonAncestor(root3, p3, q3);
        System.out.println("Test Case 2: ");
        System.out.println(solution3.val);

        //Test Case where root is LCA aka where either p/q are the root.
        TreeNode root1 = new TreeNode(200);
        //TreeNode p1 = root1.left = new TreeNode(5);
        TreeNode q1 = root1.right = new TreeNode(1);
        
        solution test2 = new solution();
        
        TreeNode solution2 = test2.lowestCommonAncestor(root1, root1, q1);
        System.out.println("Test Case 3: ");
        System.out.println(solution2.val);

        
        


        
       


     }


    
}

File: HW/HW9/solution.py
class TreeNode(object):

    def __init__(self, val=0, left=None, right=None):

        self.val = val

        self.left = left

        self.right = right
    
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if((p.val < root.val and q.val > root.val) or (p.val > root.val and q.val < root.val)):
            return root
        
        if(root.val == p.val or root.val == q.val):
            return root
        
        if(q.val < root.val and p.val < root.val):
            return self.lowestCommonAncestor(root.left,p,q)
        
        elif (q.val > root.val and p.val > root.val):
            return self.lowestCommonAncestor(root.right, p, q)


# Test Case 1: General case
root = TreeNode(4)
p = root.left = TreeNode(3)
root.right = TreeNode(8)
q = root.left.left = TreeNode(1)
root.right.left = TreeNode(5)
root.right.right = TreeNode(9)

test1 = TreeNode()
solution1 = test1.lowestCommonAncestor(root, p, q)
print("Test Case 1: ", solution1.val)

# Test Case 2: p and q are directly under the root
root2 = TreeNode(4)
p2 = root2.left = TreeNode(5)
q2 = root2.right = TreeNode(1)

test2 = TreeNode()
solution2 = test2.lowestCommonAncestor(root2, p2, q2)
print("Test Case 2: ", solution2.val)

# Test Case 3: root is LCA, where either p or q are the root
root3 = TreeNode(200)
q3 = root3.right = TreeNode(1)

test3 = TreeNode()
solution3 = test3.lowestCommonAncestor(root3, root3, q3)
print("Test Case 3: ", solution3.val)

